src/Analyzers/CSharp/CodeFixes/UseCollectionInitializer/CSharpUseCollectionInitializerCodeFixProvider_CollectionExpression.cs

```diff
@@ -0,0 +1,589 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CSharp.Analyzers.UseCollectionExpression;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.UseCollectionInitializer;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCollectionInitializer
+{
+    using static SyntaxFactory;
+
+    internal partial class CSharpUseCollectionInitializerCodeFixProvider
+    {
+        /// <summary>
+        /// Creates the final collection-expression <c>[...]</c> that will replace the given <paramref
+        /// name="objectCreation"/> expression.
+        /// </summary>
+        private static async Task<CollectionExpressionSyntax> CreateCollectionExpressionAsync(
```

Comment:
this logic is effectively entirely new.  The code heavily emphasizes trying to match the new collection-expression style against the current code the user has (or infer what would be best for certain situations).  I've tried to doc it heavily as to what it's doing.